### PR TITLE
executor: implement set transaction read only as of transaction

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/21094

### What is changed and how it works?
Implement `Set Transaction read only as of .....` statement.

### Check List 

Tests 

- Unit test

### Release note 

* No release note

### CODE REVIEW

File: executor/stale_txn_test.go [hunk 1]
```diff
@@ -267,3 +267,16 @@ func (s *testStaleTxnSerialSuite) TestStalenessTransactionSchemaVer(c *C) {
 	// got an old infoSchema
 	c.Assert(schemaVer3, Equals, schemaVer1)
 }
+
+func (s *testStaleTxnSerialSuite) TestSetTransactionReadOnlyAsOf(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec(`SET TRANSACTION READ ONLY as of timestamp '2021-04-21 00:42:12'`)
```
Comment (nolouch) [inline executor/stale_txn_test.go:273]: Should support expression? our bound read is a function call.

Comment (Yisaer) [reply executor/stale_txn_test.go]: It is already supported, the testcase updated.

File: sessionctx/variable/sysvar.go [hunk 2]
```diff
@@ -748,6 +748,13 @@ var defaultSysVars = []*SysVar{
 		}
 		return nil
 	}},
+	{Scope: ScopeSession, Name: TiDBTxnReadTS, Value: "", SetSession: func(s *SessionVars, val string) error {
+		err := setTxnReadTS(s, val)
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (morgo) [inline sessionctx/variable/sysvar.go:756]:
```suggestion
		return setTxnReadTS(s, val)
```

File: sessionctx/variable/varsutil.go [hunk 3]
```diff
@@ -465,6 +465,23 @@ func setSnapshotTS(s *SessionVars, sVal string) error {
...
+	t1, err := t.GoTime(s.TimeZone)
+	if err != nil {
+		return err
+	}
+	s.TxnReadTS = GoTimeToTS(t1)
+	return err
+}
+
 // GoTimeToTS converts a Go time to uint64 timestamp.
 func GoTimeToTS(t time.Time) uint64 {
```
Comment (JmPotato) [inline sessionctx/variable/varsutil.go:486]: I think this function is duplicated with `oracle.GoTimeToTS`, maybe we should remove it and use the latter.

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

File: sessionctx/variable/varsutil.go (hunk 3 above)
Comment (Yisaer) [reply sessionctx/variable/varsutil.go:487]: updated.

**Review by JmPotato — APPROVED** [review]

**Review by djshow832 — COMMENTED** [review]
Does `set transaction` only affect explicit transactions? Does it affect auto-commit transactions?

File: executor/simple.go [hunk 4]
```diff
@@ -566,6 +566,17 @@ func (e *SimpleExec) executeUse(s *ast.UseStmt) error {
...
+	// When TxnReadTS is not 0, it indicates the transaction is staleness transaction
+	if e.ctx.GetSessionVars().TxnReadTS > 0 {
+		startTS := e.ctx.GetSessionVars().TxnReadTS
+		// clear TxnReadTS after we used it.
+		e.ctx.GetSessionVars().TxnReadTS = 0
+		if err := e.ctx.NewTxnWithStartTS(ctx, startTS); err != nil {
+			return err
+		}
+		e.ctx.GetSessionVars().SetInTxn(true)
+		return nil
```
Comment (djshow832) [inline executor/simple.go:578]: Why do you set it and return? To walk around `s.ReadOnly` part?

File: sessionctx/variable/sysvar.go [hunk 5]
```diff
@@ -825,6 +825,13 @@ var defaultSysVars = []*SysVar{
 	}, GetSession: func(s *SessionVars) (string, error) {
 		return s.TxnScope.GetVarValue(), nil
 	}},
+	{Scope: ScopeSession, Name: TiDBTxnReadTS, Value: "", SetSession: func(s *SessionVars, val string) error {
```
Comment (djshow832) [inline sessionctx/variable/sysvar.go:828]: Is this sysvar exposed to users?

Comment (Yisaer) [reply sessionctx/variable/sysvar.go:828]: I think not

File: executor/simple.go (hunk 4 above)
Comment (Yisaer) [reply executor/simple.go]: updated.

File: sessionctx/variable/sysvar.go (hunk 5 above)
Comment (djshow832) [reply sessionctx/variable/sysvar.go:828]: The problem is, users may also use this variable to set read ts.

File: executor/simple.go [hunk 6]
```diff
@@ -584,18 +588,29 @@ func (e *SimpleExec) executeBegin(ctx context.Context, s *ast.BeginStmt) error {
...
 	// If the last un-committed/un-rollback transaction is a time-bounded read-only transaction, we should
-	// always create a new transaction.
 	txnCtx := e.ctx.GetSessionVars().TxnCtx
 	if txnCtx.History != nil || txnCtx.IsStaleness {
 		err := e.ctx.NewTxn(ctx)
 		if err != nil {
 			return err
 		}
-	}
+	} // always create a new transaction.
```
Comment (djshow832) [inline executor/simple.go]: Wrong place

**Pull request merged**